deepcopy() of Objects should call __g/setstate__

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39500 deepcopy() of Objects should call __g/setstate__**

The presence of a `__getstate__`/`__setstate__` pair on an Object indicates that the user has explicitly defined the serialization/deserialization behavior, and it is thus *not* valid to do a simple slot-wise deepcopy. We should delegate to these methods to instantiate a new object.

Differential Revision: [D21875091](https://our.internmc.facebook.com/intern/diff/D21875091)

Commit:
deepcopy() of Objects should call __g/setstate__

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -516,18 +517,28 @@ IValue IValue::deepcopy(
-      copy = IValue(toObject()->deepcopy(memo));
-      break;
+      auto class_type = type()->expect<ClassType>();
+      if (class_type->hasMethod("__getstate__") && class_type->hasMethod("__setstate__")) {
```
Ideally all this logic would live in ivalue::Object::deepcopy, but it's difficult to recover an `IValue` instance of the object once `this` has already been unwrapped to `ivalue::Object` (intrusive_ptr ownership and constness issues). So for now this just lives here

Commit:
Update on "deepcopy() of Objects should call __g/setstate__"


The presence of a `__getstate__`/`__setstate__` pair on an Object indicates that the user has explicitly defined the serialization/deserialization behavior, and it is thus *not* valid to do a simple slot-wise deepcopy. We should delegate to these methods to instantiate a new object.

Differential Revision: [D21875091](https://our.internmc.facebook.com/intern/diff/D21875091)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -516,18 +517,28 @@ IValue IValue::deepcopy(
     case IValue::Tag::Object: {
-      copy = IValue(toObject()->deepcopy(memo));
-      break;
+      auto class_type = type()->expect<ClassType>();
```
what about `Capsule`?

jamesr66a replied:
Capsule is only used in torchbind and should never appear here, as all Capsule instances will be attributes of an Object with `__getstate__`/`__setstate__`

Commit:
Update on "deepcopy() of Objects should call __g/setstate__"


The presence of a `__getstate__`/`__setstate__` pair on an Object indicates that the user has explicitly defined the serialization/deserialization behavior, and it is thus *not* valid to do a simple slot-wise deepcopy. We should delegate to these methods to instantiate a new object.

Differential Revision: [D21875091](https://our.internmc.facebook.com/intern/diff/D21875091)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LG, thanks!
